docs/exceptions.md

```diff
@@ -0,0 +1,246 @@
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* return `Either Text`
+
+Do:
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow`
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
+
+* `Maybe`, `Either e`
+* `MaybeT`, `ExceptT e`
+* `CatchT`
+
+Avoid using `Text` with the error message in place of `e` -- create a proper
+ADT. In case creating a proper ADT feels too cumbersome, use `CatchT`, which
+is equivalent to `ExceptT SomeException`.
+
+Be careful not to use `MaybeT`, `ExceptT`, and `CatchT` in potentially impure
+code. When in doubt whether the code is potentially impure, use `MonadThrow`.  
+(The reason we don't want `ExceptT` and co. in potentially impure code is that
+they add additional exception mechanisms to the one that `IO` has, and
+`catch`/`bracket` don't account for this).
+
+### Impure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `ExceptT`, `MaybeT`, or `CatchT`
```

Comment:
I am still not convinced that using `ExceptT` here is a good idea. I believe that it can be safe when used carefully, but:

* now you don't have `MonadMask` instance, meaning that you might need to rewrite code if it needs to allocate resources in the future

* now you have to lift through an additional transformer, which might contradict the practices we're establishing in CSL-1823

* the rules for exception usage become more nuanced: is it ok to do this if it's in a helper function? what if I move this helper function to top-level? what if I put it into a separate module for reuse?

I've seen your comment that you want to allow usage like this, but I don't think that it pulls its weight. Are you sure that doing `throwM` and `try` wouldn't give you the same thing?

The transformation is mechanical:

```
        fmap convertRes . runExceptT $ do
-- becomes
        fmap convertRes . try @SyncStatus $ do
```

```
                True -> throwError SSDoingRecovery
-- becomes
                True -> throwM SSDoingRecovery
```

Adding an `Exception` instance does not sound bad to me, it's derived by GHC so you don't have to write any code.

It looks like we get the same benefits without any disadvantages if we `throwM` here.

Reply:
> now you don't have `MonadMask` instance, meaning that you might need to rewrite code if it needs to allocate resources in the future

I still have `MonadMask` instance, I can just use `lift` where I need it. The only place where I can't use `lift` is `throwError`, but I don't need `MonadMask` there.

> now you have to lift through an additional transformer, which might contradict the practices we're establishing in CSL-1823

It's hard to talk about it until we come to a conclusion about CSL-1823, but I am already a bit concerned about that. If we are going to prohibit all lifting, what will we do with our tests which live in `PropertyM`? There we use `lift` in many places. AFAIU, `ExceptT` case is the same. Maybe it should be discussed in CSL-1823, not here. Maybe I am just missing something.

> the rules for exception usage become more nuanced: is it ok to do this if it's in a helper function? what if I move this helper function to top-level? what if I put it into a separate module for reuse?

It's not hard to come up with concrete rules. For example, we can prohibit `ExceptT` in types of all top-level impure or potentially impure functions. If you have a nested function which has `ExceptT` in its type and you want to make it top-level, you can change its type to `m (Either Patak Sepulka)` and convert it to `ExceptT` where it's used (which is straightforward).

> It looks like we get the same benefits without any disadvantages if we `throwM` here.

Yes, I agree that there no obvious disadvantages in using `throwM`. I never had good arguments why `ExceptT` is preferable in this case. It's mostly based on my personal feeling/intuition, rather than some fact that make `ExceptT` definite winner.
1. For me `SyncStatus` conceptually is not an exception, that's why I find it a bit counterintuitive to throw it as runtime exception. And defining its `Exception` instance. It's not about writing extra code, it's just about intuition behind this. `ExceptT` name also suggests something exceptional, but I treat it simply as a transformer which provides multiple exit points.
2. `ExceptT` version is basically just sugar for the version with multiple `if`s/`case`s. And the version with multiple `case`s is exactly how we want the code to behave. It's really clear how it works if you desugar it. On the other hand, version with `throwM` involves real runtime exceptions which are implemented _somehow_.

Anyway, as I said, I don't have and never had solid arguments in favour of `ExceptT` in this case, it's mostly my personal preference. I can easily discard this preference if others prefer `throwM` version in this case.

Reply:
Btw, with `throwM` approach someone can make joke and throw `SSInFuture` from e. g. `DB.getTipHeader` function and we'll catch it and return `SSInFuture`, though perhaps we shouldn't catch it.
It's something exotic and nobody will do it of course, so you can ignore it, just something came into my mind :)

Reply:
> I still have `MonadMask` instance, I can just use lift where I need it. 

Not quite, you can't do something like this:

```
runExceptT $ do
  doStuff
  bracket ... $ do
    if b
      then throwError e -- won't work
      else return x
```

> If we are going to prohibit all lifting, what will we do with our tests which live in PropertyM?

I believe we can use it as our base monad, because the only transformer on top that we want is `ReaderT` and I am since adding `ReaderT` is equivalent to adding a parameter to a function, I believe it should be always possible.

>  If you have a nested function which has ExceptT in its type and you want to make it top-level, you can change its type to m (Either Patak Sepulka)

But returning `m (Either Patak Sepulka)` isn't recommended either, you should return `m Sepulka` and throw `Patak`. The reason is that you wouldn't want to deal with the possibility that `Patak` might be *both* thrown and returned in `Either`. Perhaps we can invent a rule that it's OK to return something as `Left` if and only if it doesn't have an `Exception` instance (preferably, it should have an instance like `instance TypeError "NOT AN EXC" => Exception Patak` to avoid accidents).

> It's mostly based on my personal feeling/intuition

I share your feelings, I see value in using `ExceptT` in this case too. But I'm having a hard time to come up with good, simple and precise rules, to detect these cases, so let's settle for a more uniform (but perhaps not as nice) solution.

Reply:
It sounds quite sad. So AFAIU in some cases we have to use exceptions for control flow, which is generally considered anti-pattern, at least in some languages. Maybe it's not a big problem for Haskell, though. Or we can write deeply nested code, but it's also sad.
I am not saying that we shouldn't do it, I just don't like this world :/

Regarding your points:
> Not quite, you can't do something like this:

Well, there is a workaround to return `Either` from inside `bracket` and then have `ExceptT $ bracket …`. If it's too unpleasant, you can use exceptions instead. Permitting `ExceptT` in such cases doesn't mean that one must always use `ExceptT`. It's just another possible approach (which will sometimes be useful).

> I believe we can use it as our base monad, because the only transformer on top that we want is `ReaderT` and since adding `ReaderT` is equivalent to adding a parameter to a function, I believe it should be always possible.

As I wrote somewhere, having `PropertyM` in base monad will leave us without `MonadMask` instance, which is probably too bad. Anyway, perhaps we shouldn't continue discussing it here.

> But returning m (Either Patak Sepulka) isn't recommended either, you should return m Sepulka and throw Patak…

1. I think we can prohibit returning `Either` from potentially impure functions. We can always define an isomorphic ADT. E. g. we have `verifyBlocksPrefix :: OldestFirst NE Block -> m (Either VerifyBlocksException (OldestFirst NE Undo, PollModifier))` and we can change its return type to `m VerifyBlocksRes` where `data VerifyBlocksRes = VerifyBlocksSuccess (OldestFirst NE Undo, PollModifier) | VerifyBlocksInvalid VerifyBlocksException`. But maybe it would be too strict restriction. `Either` doesn't always carry something exceptional as its `Left`. For instance, we have `type Block = Either GenesisBlock  MainBlock`. This rule (the version without «it's OK to return something as Left if and only if it doesn't have» part) would mean we can't return `m Block`. Yes, we can and should change it do another ADT instead of using `Either`. But do we want to create an ADT for everything isomorphic to `Either`?
Also, let's consider `VerifyBlocksRes` from above. `VerifyBlocksException` might have an `Exception` instance. Should we care about possibility to `throwM` it? It's not `Either`, but isomorphic. What if `VerifyBlocksRes` has 10 constructors only one of which has an `Exception` instance?
Alternatively we can change `verifyBlocksPrefix`'s type to `OldestFirst NE Block -> m (OldestFirst NE Undo, PollModifier)`. My intuition suggests that we shouldn't do it, but maybe it's another case where I should ignore intuition.
2. I like this part:
> Perhaps we can invent a rule that it's OK to return something as Left if and only if it doesn't have an Exception instance (preferably, it should have an instance like instance TypeError "NOT AN EXC" => Exception Patak to avoid accidents).

3. Exceptions Best Practices suggests that returning `Either` is ok. Maybe not ideal but not something to prohibit. At least that's how I treat this phrase:

> Generally the solution to the ExceptT IO anti-pattern is to return an Either from more functions and throw an exception for uncommon errors.

Reply:
Does anyone else have an opinion on this? Should we add a rule about allowing local use of `ExceptT` in impure code? I do not have a strong opinion, but I'm hesitant to add more complications to the rules.

Reply:
I like the rule to use `ExceptT` only with `TypeError` exception instances for types. I'm not aware of better solution for _early exit_ from function (I'm aware only of `Cont`-monad solution but I can't call this solution a better one ;) ).